swap/src/alice/swap.rs

```diff
@@ -253,288 +253,292 @@ impl Swap {
-                            signed_tx_punish,
-                            self.bitcoin_wallet.clone(),
-                            self.config,
-                        );
+                self.db
+                    .insert_latest_state(self.swap_id, state::Swap::Alice((&state).into()))
+                    .await?;
 
-                        let bitcoin_wallet_clone = self.bitcoin_wallet.clone();
-
-                        let refund_tx_seen =
-                            bitcoin_wallet_clone.watch_for_raw_transaction(tx_refund.txid());
-
-                        pin_mut!(punish_tx_finalised);
-                        pin_mut!(refund_tx_seen);
-
-                        state = match select(punish_tx_finalised, refund_tx_seen).await {
-                            Either::Left(_) => AliceState::Punished,
-                            Either::Right((published_refund_tx, _)) => {
-                                let spend_key = extract_monero_private_key(
-                                    published_refund_tx,
-                                    tx_refund,
-                                    state3.s_a,
-                                    state3.a.clone(),
-                                    state3.S_b_bitcoin,
-                                )?;
-                                AliceState::BtcRefunded { spend_key, state3 }
-                            }
-                        };
-                        self.db
-                            .insert_latest_state(self.swap_id, state::Swap::Alice((&state).into()))
-                            .await?;
-                    }
-                    AliceState::XmrRefunded => {
-                        break Ok(AliceState::XmrRefunded);
-                    }
-                    AliceState::BtcRedeemed => {
-                        break Ok(AliceState::BtcRedeemed);
-                    }
-                    AliceState::Punished => {
-                        break Ok(AliceState::Punished);
-                    }
-                    AliceState::SafelyAborted => {
-                        break Ok(AliceState::SafelyAborted);
-                    }
-                }
+                Ok(state)
+            }
+            AliceState::XmrRefunded => Ok(AliceState::XmrRefunded),
+            AliceState::BtcRedeemed => Ok(AliceState::BtcRedeemed),
+            AliceState::Punished => Ok(AliceState::Punished),
+            AliceState::SafelyAborted => Ok(AliceState::SafelyAborted),
+        }
+    }
+
+    // State machine driver for swap execution
+    pub async fn run_until(
+        mut self,
+        state: AliceState,
+        is_target_state: fn(&AliceState) -> bool,
+    ) -> Result<AliceState> {
+        let mut result = state;
+        loop {
+            info!("Current state: {}", result);
+            result = self.next_state(result).await?;
+            if is_target_state(&result) {
+                return Ok(result);
```

Comment:
```
	break Ok(result);
```

`break` should be used to return a value from a loop, not `return`: https://doc.rust-lang.org/rust-by-example/flow_control/loop/return.html